keel-clouddriver/src/main/kotlin/com/netflix/spinnaker/keel/clouddriver/MemoryCloudDriverCache.kt

```diff
@@ -50,158 +39,176 @@ import java.util.function.BiFunction
  */
 class MemoryCloudDriverCache(
   private val cloudDriver: CloudDriverService,
-  private val meterRegistry: MeterRegistry,
-  private val cacheProperties: CacheProperties = CacheProperties()
+  cacheFactory: CacheFactory
 ) : CloudDriverCache {
 
-  private val securityGroupSummariesByIdOrName = buildAsyncCache<String, SecurityGroupSummary>(
-    cacheName = "securityGroups",
-    defaultExpireAfterWrite = Duration.ofMinutes(10)
-  )
+  private val securityGroupsById: AsyncLoadingCache<Triple<String, String, String>, SecurityGroupSummary> = cacheFactory
+    .asyncLoadingCache(
+      cacheName = "securityGroupsById",
+      defaultExpireAfterWrite = Duration.ofMinutes(10)
+    ) { (account, region, id) ->
+      runCatching {
+        val credential = credentialBy(account)
+        cloudDriver.getSecurityGroupSummaryById(account, credential.type, region, id, DEFAULT_SERVICE_ACCOUNT)
+          .also {
+            securityGroupsByName.synchronous().put(Triple(account, region, it.name), it)
+          }
+      }
+        .handleNotFound()
+        ?: notFound("Security group with id $id not found in the $account account and $region region")
+    }
```

Comment:
I remodeled the caches so that the cache loading logic is encapsulated in the declaration. I think this makes it easier to avoid any ad-hockery at the place the cache is used.